File: src/Pipeline.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Options;
use Logger;
use Align;
use IndexReference;
use Illumina2Sanger;
use OutputDir;
use Align2Sam;
use Sam2BamSorted;
use MergeBamsAndIndex;
use Pileup;
use VarFilter;
use Timer;

$| = 1;

Timer::init_timer('Timer.log');

Timer::timer('Start of job');

# Process command line arguments. Remove any arguments from command line vector.
my %OPTIONS = Options::get_options();

# Pick reference file and sequences from command line arguments vector.
my ($REFERENCE, @SEQUENCES) = @ARGV;

# Open the log file and redirect output to it
Logger::init_logger($OPTIONS{l});

# Step 1. Create reference database.
IndexReference::mkRefDataBase($REFERENCE, "-a $OPTIONS{i}");

Timer::timer('Reference database created');

# Step 2. Index the reference.
IndexReference::indexReference($REFERENCE);

Timer::timer('Reference database indexed');

# Create the output directory
my $DIR = OutputDir::mk_output_dir($SEQUENCES[0], $OPTIONS{d});

my $COUNT = 1;
my @SORTED_BAMS = ();
foreach (@SEQUENCES)
{
   # Step 3. Convert sequence to fastq format.
   my $SEQ_FASTQ = Illumina2Sanger::convertOneFile($_);

   Timer::timer('Sequence '.$COUNT.' converted to fastq format');

   # Step 4. Align sequence to the reference database.
   my $SEQ_ALIGN = Align::align($OPTIONS{a}, $OPTIONS{t}, $DIR, $REFERENCE, $SEQ_FASTQ);

   Timer::timer('Sequence '.$COUNT.' aligned to database');

   # Step 5. Convert alignment to SAM format.
   my $SEQ_SAM = Align2Sam::align2Sam("samse", $REFERENCE, $_, $SEQ_ALIGN);

   Timer::timer('Alignment for sequence '.$COUNT.' converted to SAM format');

   # Step 6. Convert SAM to BAM. 
   my $SEQ_BAM = Sam2BamSorted::sam2Bam("-b", $REFERENCE, $SEQ_SAM);

   Timer::timer('SAM file for sequence '.$COUNT.' converted to BAM format');

   # Step 7. Sort the BAM file.
   my $SEQ_BAM_SORTED = Sam2BamSorted::sortBam($SEQ_BAM);

   Timer::timer('BAM file for sequence '.$COUNT.' sorted');

   push @SORTED_BAMS, $SEQ_BAM_SORTED;
   $COUNT += 1;
}

# Steps 8-10. Merge BAM files, and align the result.
my $BAM_ALIGN = MergeBamsAndIndex::mergeBamsAndIndex($DIR, @SORTED_BAMS); 

Timer::timer('BAMs merged and indexed');

# Step 11. Construct pileup file. 
my $PILEUP = Pileup::pileup($DIR, "-c", $REFERENCE, $BAM_ALIGN);

Timer::timer('Pileup file created');

# Step 12-13. Run variation filter.
VarFilter::varFilter("-p -D $OPTIONS{D}", $PILEUP, ".all.snps");

Timer::timer('First VarFilter run');

VarFilter::varFilter("-D $OPTIONS{D}", $PILEUP, ".snps");

Timer::timer('Second VarFilter run');

Timer::stop_timer();

File: src/python/Pipeline.py
from os.path import (split, splitext, join, exists)
from os import (system, remove, mkdir)
from Logger import (logInfo, logWarn)
from sys import exit
from shutil import copy

def mkOutputDir(dir):
    def tryMakeDir(dir):
       if not exists(dir):
          logInfo('Creating folder %s' % dir)
          try:
              mkdir(dir, '0777')
          except IOError, e:
              exit('%s\nFailed to make directory %s' % (e, dir))
    tryMakeDir(dir)
    tryMakeDir(os.path.join(dir, 'Binary'))

def runCommand(msg, cmd):
    logInfo(msg)
    logInfo(command)
    status = system(command)
    if status != 0:
        logWarn("command '%s' returned non-zero status: %d'" % (command, status))
    return status

def align(algorithm, threadFlags, dir, reference, sequence):
    (root, ext) = splitext(sequence)
    if ext != ".fastq":
        exit("align: sequence file %s does not have .fastq extension" % sequence)
    alignmentFile = root + ".sai"
    command = "bwa %s -t %s %s %s > %s" % (algorithm, threadFlags, reference, sequence, alignmentFile)
    runCommand ("Running Alignment", command)
    return alignment

def align2Sam (flags, reference, sequence, alignment):
    (root, ext) = splitext(alignment)
    if ext != ".sai":
        exit("align2Sam: alignment file %s does not have .sai extension" % alignment)
    samfile = root + ".sam"
    command = "bwa %s %s %s %s > %s" % (flags, reference, alignment, sequence, samfile)
    runCommand("Converting alignment to SAM format", command)
    return samfile

def ills2FastQs(sequences):
    return map(illumina2FastQ, sequences)

def illumina2FastQ(file):
    (root, ext) = split(file)
    sangerFile = root + ".fastq"
    if ext == ".fastq":
        logInfo("illumina2FastQ: file %s already in Standard/Sanger FASTQ Format" % file)
    elif ext == ".txt":
        command = "maq ill2sanger %s %s" % (file, sangerFile)
        runCommand("Converting illumina file to Standard/Sanger FASTQ Format", command)
    else:
        exit("sequence file %s not in illumina (.txt) or sanger (.fastq) format" % file)
    return sangerFile

# create reference database
def mkRefDataBase(reference, indexFlags):
    refDatabase = referenece + ".bwt"
    if exists(refDatabase) 
        logInfo("Reference database already exists: using %s" % refDatabase)
    else:
        command = "bwa index %s %s" % (indexFlags, reference)
        runCommand("Creating Reference Database", command)
    return refDatabase 

# index the reference
def indexReference(reference):
    referenceIndex = reference + .".fai"
    if exists(referenceIndex):
        logInfo("Reflist already exists: using %s" % referenceIndex)
    else:
        command = "samtools faidx %s" % reference
        runCommand("Creating Reflist", command)
    return referenceIndex

def mergeBamsAndIndex(dir, bamFiles):
    bamAlignFile = dir + "Binary/all_reads_aligned.bam"
    numBams = len(bamFiles);
    # If there is more than one BAM file then merge them.
    if numBams > 1:
        command = "samtools merge %s %s" % (bamAlignFile, ' '.join(bamFiles))
        runCommand ("Merging Bam files", command)
    elif numBams == 1:
        copy(bamfiles[0], bamAlignFile)
    else:
        exit("Attempt to merge and index zero BAM files")
    command = "samtools index %s" % bamAlignFile
    runCommand("Indexing Alignment File", command)
    return bamAlignFile

def pileup(dir, pileupFlags, reference, bamAlignFile):
    pileupFile = dir + "reads_aligned.pileup"
    command = "samtools pileup %s -f %s %s > %s" % (pileupFlags, reference, bamAlignFile, pileupFile)
    runCommand("Constructing Pileup File", command)
    return pileupFile

# Convert SAM to BAM
# Assumes the Binary directory has been created already.
def sam2Bam(vregion, viewFlag, reference, samFile):
    (root, base) = split(samFile)
    (name, ext) = splitext(base)
    if ext != ".sam":
        exit("sam2Bam: file %s does not end in .sam extension" % samFile)
    bamAlignFile = os.path.join(root + "Binary", name + ".bam")
    command = "samtools view $VIEWFLAG -t $REFERENCE.fai -o $BAMALIGN $SAMFILE @VREGION";
    faiFile = reference + ".fai"
    command = "samtools view %s -t %s -o %s %s %s" % (viewFlag, faiFile, bamAlignFile, samFile, vregion) 
    runCommand("Converting to BAM", command)
    return bamAlignFile

# Sort a BAM file
def sortBam(bamAlignFile, sortFlags):
    (root, ext) = splitext(bamAlignFile)
    sortedBamFile = root + ".sorted"
    command = "samtools sort %s %s %s" % (sortFlags, bamAlignFile, sortedBamFile)
    runCommand("Sorting Bam alignments", command)
    # I think we should leave the decision to remove the BAM file to the
    # caller of this function.
    #remove(bamAlignFile)
    return sortedBamFile + ".bam"

# run the variations filter
def varFilter(varFilterFlags, pileupFile, outputExtension):
   snpsFile = pileupFile + outputExtension
   command = "samtools.pl varFilter $VARFILTERFLAGS $PILEUPFILE &> $SNPSFILE";
   command = "samtools.pl varFilter %s %s &> %s" % (varFilterFlags, pileupFile, snpsFile)
   runCommand("Running Variations Filter", command)
   return snpsFile 
